Added naming convention details for C# code (#24054)
dotnet/docs: docs/csharp/programming-guide/inside-a-program/coding-conventions.md

---
title: "C# Coding Conventions - C# Programming Guide"
description: Learn about coding conventions in C#. Coding conventions create a consistent look to the code and facilitate copying, changing, and maintaining the code.
ms.date: 05/05/2021
helpviewer_keywords:
  - "coding conventions, C#"
  - "Visual C#, coding conventions"
  - "C# language, coding conventions"
---

# C# Coding Conventions (C# Programming Guide)

Coding conventions serve the following purposes:

> - They create a consistent look to the code, so that readers can focus on content, not layout.
> - They enable readers to understand the code more quickly by making assumptions based on previous experience.
> - They facilitate copying, changing, and maintaining the code.
> - They demonstrate C# best practices.

Author: ```suggestion
> [!div class="checklist"]
>
> - They create a consistent look to the code, so that readers can focus on content, not layout.
> - They enable readers to understand the code more quickly by making assumptions based on previous experience.
> - They facilitate copying, changing, and maintaining the code.
> - They demonstrate C# best practices.
```

The guidelines in this article are used by Microsoft to develop samples and documentation.

## Naming conventions

There are several naming conventions to consider when writing C# code.

### Pascal case

You should use pascal casing ("PascalCasing") when naming a `class`, `record`, or `struct`.

```csharp
public class DataService
{
}
```

```csharp
public record PhysicalAddress(
    string Street,
    string City,
    string StateOrProvince,
    string ZipCode);
```

```csharp
public struct ValueCoordinate
{
}
```

When naming an `inteface`, use pascal casing in addition to prefixing the name with an `I`. This clearly indicates to consumers that it's an `interface`.

```csharp
public interface IWorkerQueue
{
}
```

> When naming `public` members of objects, such as; fields, properties, events, methods, and local functions use pascal casing.

Reviewer: nit:

```suggestion
When naming `public` members of types, such as; fields, properties, events, methods, and local functions use pascal casing.
```

---
title: "C# Coding Conventions - C# Programming Guide"
description: Learn about coding conventions in C#. Coding conventions create a consistent look to the code and facilitate copying, changing, and maintaining the code.
ms.date: 05/05/2021
helpviewer_keywords:
  - "coding conventions, C#"
  - "Visual C#, coding conventions"
  - "C# language, coding conventions"
---

# C# Coding Conventions (C# Programming Guide)

Coding conventions serve the following purposes:

> [!div class="checklist"]
>
> - They create a consistent look to the code, so that readers can focus on content, not layout.
> - They enable readers to understand the code more quickly by making assumptions based on previous experience.
> - They facilitate copying, changing, and maintaining the code.
> - They demonstrate C# best practices.

The guidelines in this article are used by Microsoft to develop samples and documentation.

## Naming conventions

There are several naming conventions to consider when writing C# code.

### Pascal case

> You should use pascal casing ("PascalCasing") when naming a `class`, `record`, or `struct`.

Reviewer 2: ```suggestion
Use pascal casing ("PascalCasing") when naming a `class`, `record`, or `struct`.
```

```csharp
public class DataService
{
}
```

```csharp
public record PhysicalAddress(
    string Street,
    string City,
    string StateOrProvince,
    string ZipCode);
```

```csharp
public struct ValueCoordinate
{
}
```

When naming an `inteface`, use pascal casing in addition to prefixing the name with an `I`. This clearly indicates to consumers that it's an `interface`.

```csharp
public interface IWorkerQueue
{
}
```

> When naming `public` members of types, such as; fields, properties, events, methods, and local functions use pascal casing.

Reviewer 2: ```suggestion
When naming `public` members of types, such as fields, properties, events, methods, and local functions, use pascal casing.
```

```csharp
public class PartyEvents
{
    // A public field, these should be used sparingly
    public bool IsPartyTime;

    // An init-only property
    public HashSet<Attendee> Attendees { get; init; }

    // An event
    public event Action HappyHour;

    // Method
    public void StartEventProcessing()
    {
        // Local function
        static int CountAttendees() => Attendees.Count;
        // ...
    }
}
```

> When writing positional-records, use pascal casing for parameters as they're the public properties of the record.

Reviewer 2: Should positional-records by hyphenated?

```csharp
public record PhysicalAddress(
    string Street,
    string City,
    string StateOrProvince,
    string ZipCode);
```

> For more information on positional-records, see [Positional syntax for property definition](../../language-reference/builtin-types/record.md#positional-syntax-for-property-definition).

Reviewer 2: Same.

### Camel case

> You should use camel casing ("camelCasing") when naming `private` or `internal` fields, and they should be prefixed with `_`.

Reviewer 2: ```suggestion
Use camel casing ("camelCasing") when naming `private` or `internal` fields, and prefix them with `_`.
```

```csharp
public class DataService
{
    private IWorkerQueue _workerQueue;
}
```

> [!TIP]
> When editing C# code that follows these naming conventions in an IDE that supports statement completion, typing `_` will show all of the object-scoped members.

When working with `static` fields that are `private` or `internal`, use the `s_` prefix and for thread static use `t_`.

```csharp
public class DataService
{
    private static IWorkerQueue s_workerQueue;

    [ThreadStatic]
    private static TimeSpan t_timeSpan;
}
```

When writing method parameters, use camel casing.

```csharp
public T SomeMethod<T>(int someNumber, bool isValid)
{
}
```

For more information on C# naming conventions, see [C# Coding Style](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md).

### Additional naming conventions

> - Examples that don't include [using directives](../../language-reference/keywords/using-directive.md), use namespace qualifications. If you know that a namespace is imported by default in a project, you don't have to fully qualify the names from that namespace. Qualified names can be broken after a dot (.) if they are too long for a single line, as shown in the following example.

Reviewer 2: The first sentence here doesn't make sense to me.
